IAB Ran tutorial (#11)
open-5g/openrangym: _posts/tutorials/2022-02-01-iab-tutorial.md

<!-- Later on, we'll describe how to deploy a realistic IAB network using the scenarios we developed using real data from different European cities. -->

## IAB RAN

#### Prerequisites

In the remaining of this tutorial, we assume users started an interactive reservation on Colosseum with at least 4 IAB RAN nodes (`oai-iab-ran`), one OAI IAB-aware core network (`oai-core-iab`) to deploy the network. The IAB network is then controlled by [iab-manager](https://github.com/wineslab/iab-manager), that can be run either locally or in the core network node.

### Linear IAB Network

Once the reservation is ready, iab-manager can be used to interact with the different components. First of all, if you haven't already downloaded it, clone the iab-manager repo on your local machine:

{% highlight bash %}
git clone https://github.com/wineslab/iab-manager.git
{% endhighlight %}

> Then enter the directory and create a `.env` in the main folder containing your colosseum credentials in the format:

Reviewer: ```suggestion
Then enter the directory and create a `.env` in the main folder containing your Colosseum credentials in the format:
```

Then, you must configure iab-manager through the `config.yml` configuration file. Create the file with the following content:

{% highlight yaml %}
topology: "toy1.graphml"
scenario: 20055
scenario_nodes: 25
srn_blacklist: []
results_folder: "./results"
if_freq: 0
{% endhighlight %}

You can then run iab-manager:
{% highlight bash %}
python3 iab-manager.py
{% endhighlight %}

The software will connect to the Colosseum reservation API to fetch the information about the current reservation, then it will read the content of the topology from `toy1.graphml`, which represents the linear topology we want to deploy and finally it will map each SRN from the reservation to a node from the topology, which will then be mapped to a radio in the scenario. At the end of the setup, iab-manager will show a prompt that allows the user to interact with the running network.

First, let's start the core network by typing `core start`.

Then, we can start the rf scenario by typing `rf_scenario start`.

Finally, we automatically start the network by issuing the command `autostart`.

> It is also possible to start the network manually by starting first the iab-donor, then the iab-relays and finally the UE: `donor 0 start`, `iab_node start 12`, `iab_node start 34`, `iab_node start 56`, `iab_node start 78`, `ue 9 start`

Reviewer: I think there was a comment on "what are these numbers", but I don't see it (even resolved), so I'll write it again: what are these numbers? How can a reader of this tutorial interpret (and in case) change them? Do they need to change them or not?

Author: I answered it, but it got deleted by gh.
Those numbers are the ids of the nodes in the topology file (`toy1.graphml `). There's a dynamical matching between SRNs  (colosseum), Radio IDs (MCHEM) and Topology IDs that is automatically done by iab-manager.However, I'm uncertain how deep I should go into the details of this and in general of the mechanics of iab-manager into this tutorial.

Reviewer: I see. In general, as a user, if I try to run the command with these numbers and it does not work because the IDs are not the same anymore, I will be lost unless there is some extra info. Makes sense?

Author: Why the IDs should change? The toy.topology file shouldn't be changed. Anyhow if you're worried about that, we can include it in the tutorial.

Reviewer: @gabri94

<!-- Later on, we'll describe how to deploy a realistic IAB network using the scenarios we developed using real data from different European cities. -->

## IAB RAN

#### Prerequisites

In the remaining of this tutorial, we assume users started an interactive reservation on Colosseum with at least 4 IAB RAN nodes (`oai-iab-ran`), one OAI IAB-aware core network (`oai-core-iab`) to deploy the network. For both images the login is `root` and the password is `pass`. The IAB network is then controlled by [iab-manager](https://github.com/wineslab/iab-manager), that can be run either locally or in the core network node.

### Linear IAB Network

Once the reservation is ready, iab-manager can be used to interact with the different components. First of all, if you haven't already downloaded it, clone the iab-manager repo on your local machine:

{% highlight bash %}
git clone https://github.com/wineslab/iab-manager.git
{% endhighlight %}

Then enter the directory and create a `.env` in the main folder containing your Colosseum credentials in the format:

{% highlight python %}
COLOSSEUM_USER = [username]
COLOSSEUM_PWD = [pwd]
{% endhighlight %}

> This will allow iab-manager to fetch the reservations' data from the Colosseum API and automatically map the SRNs to iab-nodes of your network.

Reviewer: ```suggestion
This will allow iab-manager to fetch the reservations' data from the Colosseum API and automatically map the SRNs to IAB-nodes of your network.
```

Then, you must configure iab-manager through the `config.yml` configuration file. Create the file with the following content:

{% highlight yaml %}
topology: "toy1.graphml"
scenario: 20055
scenario_nodes: 25
srn_blacklist: []
results_folder: "./results"
if_freq: 0
{% endhighlight %}

You can then run iab-manager:
{% highlight bash %}
python3 iab-manager.py
{% endhighlight %}

The software will connect to the Colosseum reservation API to fetch the information about the current reservation, then it will read the content of the topology from `toy1.graphml`, which represents the linear topology we want to deploy and finally it will map each SRN from the reservation to a node from the topology, which will then be mapped to a radio in the scenario. At the end of the setup, iab-manager will show a prompt that allows the user to interact with the running network.

First, let's start the core network by typing `core start`.

Then, we can start the rf scenario by typing `rf_scenario start`.

Finally, we automatically start the network by issuing the command `autostart`.

> It is also possible to start the network manually by starting first the iab-donor, then the iab-relays and finally the UE: `donor 0 start`, `iab_node start 12`, `iab_node start 34`, `iab_node start 56`, `iab_node start 78`, `ue 9 start`.

Reviewer: ```suggestion
It is also possible to start the network manually by starting first the IAB-donor, then the IAB-nodes and finally the UE: `donor 0 start`, `iab_node start 12`, `iab_node start 34`, `iab_node start 56`, `iab_node start 78`, `ue 9 start`.
```

> The numbers in the commands above identify the id of the node, which is set in the topology file (`toy1.graphml`). In the case of iab-relays, since two SRNs are combined to form a single iab-relay, the id is set in a separate property of the graph called `iab`. In this example is set to the union of two single ids, so if an iab-node is made of an MT with id `1` and a DU with id `2`, its id will be `12`.

Reviewer: ```suggestion
The numbers in the commands above identify the id of the node, which is set in the topology file (`toy1.graphml`). In the case of IAB-nodes, since two SRNs are combined to form a single IAB-node, the id is set in a separate property of the graph called `iab`. In this example is set to the union of two single ids, so if an IAB-node is made of an MT with id `1` and a DU with id `2`, its id will be `12`.
```

short-description: How to setup and use and IAB network controlled by O-RAN in Colosseum
---

In this tutorial, we show how to use OpenAirInterface (OAI) with a [custom IAB implementation](https://dl.acm.org/doi/10.1145/3495243.3558750) to set up a working IAB 5G network deployed on the [Colosseum](experimental-platforms/colosseum) wireless network emulator and controlled by the RIC. The network will be made of 10 SRN, divided into 1 IAB-donor, 4 IAB-relay (each one made of one softUE and one gNB), and 1 softUE connected to the last IAB-relay. In this way, it will be possible to assess the performances of a linear IAB network with 5 hops.

<!-- Later on, we'll describe how to deploy a realistic IAB network using the scenarios we developed using real data from different European cities. -->

## IAB RAN

#### Prerequisites

> In the remaining of this tutorial, we assume users started an interactive reservation on Colosseum with at least 4 IAB RAN nodes (`oai-iab-ran` or `oai-iab-ran-stable` image), one OAI IAB-aware core network (`oai-core-iab`) to deploy the network. The IAB network is then controlled by [iab-manager](https://github.com/wineslab/iab-manager), that can be run either locally or in the core network node.

Reviewer: What is the difference between these two images?

Reviewer: Can we keep one?

Author: I'd have a `stable` image that's kept at commit that we know it works and we rarely updates after testing it thorougly, and a `develop` image that's more often (ideally continuously) updated to be on top of develop.

Reviewer: Ok but let's point the users to the stable image

### Linear IAB Network

> Once the reservation is ready, iab-manager can be used to interact with the different components. First of all, create a `.env` in the main folder containing your colosseum credentials in the format:

Reviewer: main folder of what? Where should iab-manager go?

Author: In the main folder of iab-manager, which, as stated in L18 goes either on your laptop or on the core node.

Reviewer: Please clarify also here, thanks!

{% highlight python %}
COLOSSEUM_USER = [username]
COLOSSEUM_PWD = [pwd]
{% endhighlight %}

This will allow iab-manager to fetch the reservations' data from the Colosseum API and automatically map the SRNs to iab-nodes of your network.

> Then, you must configure `iab-manager` through the `config.yml` configuration file. Create the file with the following content:

Reviewer: Check `iab-manager` vs iab-manager

{% highlight yaml %}
topology: "toy1.graphml"
scenario: 20055
scenario_nodes: 25
srn_blacklist: []
results_folder: "./results"
if_freq: 0
{% endhighlight %}

You can then run iab-manager:
{% highlight bash %}
python3 iab-manager.py
{% endhighlight %}

The software will connect to the Colosseum reservation API to fetch the information about the current reservation, then it will read the content of the topology from `toy1.graphml`, which represents the linear topology we want to deploy and finally it will map each SRN from the reservation to a node from the topology, which will then be mapped to a radio in the scenario. At the end of the setup, iab-manager will show a prompt that allows the user to interact with the running network.

First, let's start the core network by typing `core start`.

Then, we can start the rf scenario by typing `rf_scenario start`.

Finally, we automatically start the network by issuing the command `autostart`.

> It's also possible to start the network manually by starting first the iab-donor, then the iab-relays and finally the UE: `donor 0 start`, `iab_node start 12`, `iab_node start 34`, `iab_node start 56`, `iab_node start 78`, `ue 9 start`

Reviewer: ```suggestion
It is also possible to start the network manually by starting first the iab-donor, then the iab-relays and finally the UE: `donor 0 start`, `iab_node start 12`, `iab_node start 34`, `iab_node start 56`, `iab_node start 78`, `ue 9 start`
```

Reviewer: Also, what are these numbers?

Author: Each node in the topology.graphml file has an ID that is used to identify it. I didn't wanted to go down in the details of how iab-manager works as I want to keep the tutorial simple.

<!-- ### Realistic IAB Network -->

## RIC

> 

Reviewer: ```suggestion
To be updated.

```

### Near-RT RIC setup

> 

Reviewer: ```suggestion
To be updated.

```

> ### Non-RT RIC Setup

Reviewer: ```suggestion
### Non-RT RIC Setup
To be updated.

```
